providers/default/ciphers/cipher_chacha20.c

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Dispatch functions for chacha20 cipher */
+
+#include "cipher_chacha20.h"
+#include "internal/providercommonerr.h"
+
+#define CHACHA20_KEYLEN (CHACHA_KEY_SIZE)
+#define CHACHA20_BLKLEN (1)
+#define CHACHA20_IVLEN (CHACHA_CTR_SIZE)
+#define CHACHA20_FLAGS (EVP_CIPH_CUSTOM_IV | EVP_CIPH_ALWAYS_CALL_INIT)
+
+#define chacha20_einit chacha20_init
+#define chacha20_dinit chacha20_init
+#define chacha20_gettable_params cipher_generic_gettable_params
+#define chacha20_cipher chacha20_update
+
+static OSSL_OP_cipher_newctx_fn chacha20_newctx;
+static OSSL_OP_cipher_freectx_fn chacha20_freectx;
+static OSSL_OP_cipher_dupctx_fn chacha20_dupctx;
+static OSSL_OP_cipher_encrypt_init_fn chacha20_init;
+static OSSL_OP_cipher_update_fn chacha20_update;
+static OSSL_OP_cipher_final_fn chacha20_final;
+static OSSL_OP_cipher_get_params_fn chacha20_get_params;
+static OSSL_OP_cipher_get_ctx_params_fn chacha20_get_ctx_params;
+static OSSL_OP_cipher_set_ctx_params_fn chacha20_set_ctx_params;
+static OSSL_OP_cipher_gettable_ctx_params_fn chacha20_gettable_ctx_params;
+static OSSL_OP_cipher_settable_ctx_params_fn chacha20_settable_ctx_params;
+
+int CHACHA20_init_key(void *vctx, const unsigned char user_key[CHACHA_KEY_SIZE],
+                      const unsigned char iv[CHACHA_CTR_SIZE], int enc)
+{
+    PROV_CHACHA20_CTX *ctx = (PROV_CHACHA20_CTX *)vctx;
+    unsigned int i;
+
+    if (user_key != NULL)
+        for (i = 0; i < CHACHA_KEY_SIZE; i += 4)
+            ctx->key.d[i / 4] = CHACHA_U8TOU32(user_key + i);
+
+    if (iv != NULL)
+        for (i = 0; i < CHACHA_CTR_SIZE; i += 4)
+            ctx->counter[i / 4] = CHACHA_U8TOU32(iv + i);
+
+    ctx->partial_len = 0;
+    return 1;
+}
+
+int CHACHA20_cipher(void *vctx, unsigned char *out, const unsigned char *inp,
+                    size_t len)
+{
+    PROV_CHACHA20_CTX *ctx = (PROV_CHACHA20_CTX *)vctx;
+    unsigned int n, rem, ctr32;
+
+    if ((n = ctx->partial_len)) {
+        while (len && n < CHACHA_BLK_SIZE) {
+            *out++ = *inp++ ^ ctx->buf[n++];
+            len--;
+        }
+        ctx->partial_len = n;
+
+        if (len == 0)
+            return 1;
+
+        if (n == CHACHA_BLK_SIZE) {
+            ctx->partial_len = 0;
+            ctx->counter[0]++;
+            if (ctx->counter[0] == 0)
+                ctx->counter[1]++;
```

Comment:
Why do we only do this for the first two values of counter?

Reply:
The code says it operates on a 32 bit counter..

Reply:
? That makes even less sense, since counter[0] is 32 bits right?

Reply:
It's an _unsigned int_.  We might do better declaring it as _uint32_t_ at some point in the future.

Reply:
Don't our sanity tests *require* an unsigned int to be at least 32 bits?

Reply:
I think I am not going to mess with the original code in this PR as it looks like a quirk of the chacha algorithm.
The counter gets filled with 32 bit iv values initially....
It looks like the user's responsiblilty not to let the count get that high.
The function that recieves the counter[] casts all its input to 32bit. It also increments internally by 1 in a similiar way internally if the first 32 bits overflow.
Why an unsigned int was used at the top level I do not know..

Reply:
Bernstein's implementation does something similar..
i.e: https://github.com/grigorig/chachapoly/blob/master/chacha.c

    j12 = PLUSONE(j12);
    if (!j12) {
      j13 = PLUSONE(j13);
      /* stopping at 2^70 bytes per nonce is user's responsibility */
}